src/plonk/lookup/prover.rs

```diff
@@ -0,0 +1,712 @@
+                Any::Fixed => fixed_values[table.index()].clone(),
+                Any::Aux => aux_values[table.index()].clone(),
+            })
+            .collect();
+
+        // Compressed version of table columns
+        let compressed_table_value = unpermuted_table_values
+            .iter()
+            .fold(domain.empty_lagrange(), |acc, table| acc * *theta + table);
+
+        // Permute compressed (InputColumn, TableColumn) pair
+        let (permuted_input_value, permuted_table_value) =
+            permute_column_pair::<C>(domain, &compressed_input_value, &compressed_table_value)?;
+
+        // Construct Permuted struct
+        let permuted_input_poly = pk.vk.domain.lagrange_to_coeff(permuted_input_value.clone());
+        let permuted_input_coset = pk
+            .vk
+            .domain
+            .coeff_to_extended(permuted_input_poly.clone(), Rotation::default());
+        let permuted_input_inv_coset = pk
+            .vk
+            .domain
+            .coeff_to_extended(permuted_input_poly.clone(), Rotation(-1));
+
+        let permuted_input_blind = Blind(C::Scalar::rand());
+        let permuted_input_commitment = params
+            .commit_lagrange(&permuted_input_value, permuted_input_blind)
+            .to_affine();
+
+        let permuted_table_poly = pk.vk.domain.lagrange_to_coeff(permuted_table_value.clone());
+        let permuted_table_coset = pk
+            .vk
+            .domain
+            .coeff_to_extended(permuted_table_poly.clone(), Rotation::default());
+        let permuted_table_blind = Blind(C::Scalar::rand());
+        let permuted_table_commitment = params
+            .commit_lagrange(&permuted_table_value, permuted_table_blind)
+            .to_affine();
+
+        // Hash each permuted input commitment
+        transcript
+            .absorb_point(&permuted_input_commitment)
+            .map_err(|_| Error::TranscriptError)?;
+
+        // Hash each permuted table commitment
+        transcript
+            .absorb_point(&permuted_table_commitment)
+            .map_err(|_| Error::TranscriptError)?;
+
+        Ok(Permuted {
+            permuted_input_value,
+            permuted_input_poly,
+            permuted_input_coset,
+            permuted_input_inv_coset,
+            permuted_input_blind,
+            permuted_input_commitment,
+            permuted_table_value,
+            permuted_table_poly,
+            permuted_table_coset,
+            permuted_table_blind,
+            permuted_table_commitment,
+        })
+    }
+
+    /// Given a Lookup with input columns, table columns, and the permuted
+    /// input column and permuted table column, this method constructs the
+    /// grand product polynomial over the lookup. The grand product polynomial
+    /// is used to populate the Product<C> struct. The Product<C> struct is
+    /// added to the Lookup and finally returned by the method.
+    pub(in crate::plonk) fn commit_product<
```

Comment:
I'm somewhat confused about why this method is defined here. It should be defined on `Permuted<C>` right?

Reply:
I see what's going on. You needed `unpermuted_input_values` and `unpermuted_table_values` again here and thus needed `Argument` to get them. Instead, as per my other comment, this function should be defined on `Permuted<C>` and the vectors `unpermuted_input_values` / `unpermuted_table_values` should be vectors of references to polynomials and placed in `Permuted<C>` so that you don't need `Argument` again.